mypy/typeshed/stdlib/typing.pyi

```diff
@@ -902,10 +902,10 @@ if sys.version_info >= (3, 9):
     @overload
     def get_origin(tp: GenericAlias) -> type: ...
     @overload
-    def get_origin(tp: Any) -> Any | None: ...
+    def get_origin(tp: Any) -> object: ...
```

Comment:
Looking at the [implementation](https://github.com/python/cpython/blob/v3.12.0/Lib/typing.py#L2263-L2289), I think it shouldn't be too difficult to use a couple of generic `Protocol`s here so that you can have an exact return type. I know that in practice this won't matter much since `GenericAlias` and `UnionType` and stuff aren't generic right now, but that might not be the case forever (wink wink).

Reply:
let's address that in a follow-up, for now I think `HasBool` would be better

Reply:
but `object` has no `__bool__` 🤔

Reply:
😳

Reply:
hmm, object is fine then i guess

Reply:
in case you didn't know already, this is also a thing:

```pycon
>>> class Empty:
...     def __len__(self, /) -> {0}:
...         return 0
... 
>>> bool(Empty())
False
>>> Empty.__bool__
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: type object 'Empty' has no attribute '__bool__'. Did you mean: '__doc__'?
>>> "No I meant __bool__ you dumb-ass"
```

Reply:
> ` -> {0}`

🤨

yeah i knew that one, theres a good mypy rule about it `truthy-bool`

what i think would be useful are `Truthy`/`Falsey` types:
```py
def f(t: Truthy):
    ...
f(None)  # fail, `None` can't be `True`
```

maybe also a `Booly`, that is known to implement a `__bool__` or `__len__`, so more specific than `object`, kinda like `helpful-string`

Reply:
> > ` -> {0}`
> 
> 🤨, is this `Literal` shorthand in unpy?


Yea I think it will be; because that way the functionalities of `|` and `&` emerge naturally

> yeah i knew that one, theres a good mypy rule about it `truthy-bool`
> 
> what i think would be useful are `Truthy`/`Falsey` types:
> 
> ```python
> def f(t: Truthy):
>     ...
> f(None)  # fail, `None` can't be `True`
> ```

I see what you mean, but in this case

```pycon
>>> None.__bool__()
False
```

> maybe also a `Booly`, that is known to implement a `__bool__` or `__len__`, so more specific than `object`, kinda like `helpful-string`

I like that idea, and I've played around with it myself. But the issue here is that upstream typeshed always types `__len__: () -> int`, so it's never `Literal[0]`. 

That's also the reason for these `Empty` optypes btw: https://github.com/jorenham/optype/blob/6a153d9c4dde6fefa6bce4a6c679f3e5135cb50b/optype/typing.py#L60-L83